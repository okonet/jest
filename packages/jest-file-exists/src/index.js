/**
 * Copyright (c) 2014, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */

'use strict';

import type {Path} from 'types/Config';
import type {HasteFS} from 'types/HasteMap';

const fs = require('fs');

module.exports = (
  filePath: Path,
  hasteFS: ?HasteFS,
): boolean => (hasteFS && hasteFS.exists(filePath)) || fs.existsSync(filePath);
