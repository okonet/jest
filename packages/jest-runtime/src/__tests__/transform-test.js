/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+jsinfra
 */
'use strict';

const skipOnWindows = require('skipOnWindows');

jest
  .mock('graceful-fs')
  .mock('jest-file-exists')
  .mock('jest-util')
  .mock('vm');

jest.mock(
  'test-preprocessor',
  () => {
    const escapeStrings = str => {
      return str.replace(/'/, `'`);
    };

    return {
      getCacheKey: jest.fn((content, filename, configStr) => 'ab'),
      process: (content, filename, config) => {
        return (`
          const TRANSFORMED = {
            filename: '${escapeStrings(filename)}',
            script: '${escapeStrings(content)}',
            config: '${escapeStrings(JSON.stringify(config))}',
          };
        `);
      },
    };
  },
  {virtual: true},
);

jest.mock(
  'css-preprocessor',
  () => {
    return {
      getCacheKey: jest.fn((content, filename, configStr) => 'cd'),
      process: (content, filename, config) => {
        return (`
          module.exports = {
            filename: ${filename},
            rawFirstLine: ${content.split('\n')[0]},
          };
        `);
      },
    };
  },
  {virtual: true},
);

const getCachePath = (fs, config) => {
  for (const path in mockFs) {
    if (path.startsWith(config.cacheDirectory)) {
      return path;
    }
  }
  return null;
};

let config;
let fs;
let mockFs;
let object;
let transform;
let vm;

describe('transform', () => {

  const reset = () => {
    jest.resetModules();

    object = data => Object.assign(Object.create(null), data);

    vm = require('vm');

    mockFs = object({
      '/fruits/banana.js': [
        'module.exports = "banana";',
      ].join('\n'),
      '/fruits/kiwi.js': [
        'module.exports = () => "kiwi";',
      ].join('\n'),
      '/node_modules/react.js': [
        'module.exports = "react";',
      ].join('\n'),
      '/styles/App.css': [
        'root {',
        '  font-family: Helvetica;',
        '}',
      ].join('\n'),
    });

    fs = require('graceful-fs');
    fs.readFileSync = jest.fn((path, options) => {
      expect(options).toBe('utf8');

      if (mockFs[path]) {
        return mockFs[path];
      }

      throw new Error(`Cannot read path '${path}'.`);
    });
    fs.writeFileSync = jest.fn((path, data, options) => {
      expect(options).toBe('utf8');
      mockFs[path] = data;
    });

    fs.unlinkSync = jest.fn();
    fs.statSync = jest.fn(path => ({
      isFile: () => !!mockFs[path],
      mtime: {getTime: () => 42},
    }));

    require('jest-file-exists').mockImplementation(path => !!mockFs[path]);

    config = {
      cache: true,
      cacheDirectory: '/cache/',
      name: 'test',
      transformIgnorePatterns: ['/node_modules/'],
      rootDir: '/',
    };

    transform = require('../transform');
  };

  beforeEach(reset);

  it('transforms a file properly', () => {
    config.collectCoverage = true;
    const response = transform('/fruits/banana.js', config);

    expect(response instanceof vm.Script).toBe(true);
    expect(vm.Script.mock.calls[0][0]).toMatchSnapshot();

    // no-cache case
    expect(fs.readFileSync.mock.calls.length).toBe(1);
    expect(fs.readFileSync).toBeCalledWith('/fruits/banana.js', 'utf8');

    // in-memory cache
    const response2 = transform('/fruits/banana.js', config);
    expect(response2).toBe(response);

    transform('/fruits/kiwi.js', config);
    const snapshot = vm.Script.mock.calls[1][0];
    expect(snapshot).toMatchSnapshot();

    transform(
      '/fruits/kiwi.js',
      Object.assign({}, config, {collectCoverage: true}),
    );

    expect(vm.Script.mock.calls[0][0]).not.toEqual(snapshot);
    expect(vm.Script.mock.calls[0][0]).not.toMatch(/instrumented kiwi/);

    // If we disable coverage, we get a different result.
    transform(
      '/fruits/kiwi.js',
      Object.assign({}, config, {collectCoverage: false}),
    );
    expect(vm.Script.mock.calls[1][0]).toEqual(snapshot);
  });

  it('uses the supplied preprocessor', () => {
    config = Object.assign(config, {
      transform: [['^.+\\.js$', 'test-preprocessor']],
    });

    transform('/fruits/banana.js', config);

    expect(require('test-preprocessor').getCacheKey).toBeCalled();

    expect(vm.Script.mock.calls[0][0]).toMatchSnapshot();

    transform('/node_modules/react.js', config);
    // ignores preprocessor
    expect(vm.Script.mock.calls[1][0]).toMatchSnapshot();
  });

  it('uses multiple preprocessors', () => {
    config = Object.assign(config, {
      transform: [
        ['^.+\\.js$', 'test-preprocessor'],
        ['^.+\\.css$', 'css-preprocessor'],
      ],
    });

    transform('/fruits/banana.js', config);
    transform('/styles/App.css', config);

    expect(require('test-preprocessor').getCacheKey).toBeCalled();
    expect(require('css-preprocessor').getCacheKey).toBeCalled();
    expect(vm.Script.mock.calls[0][0]).toMatchSnapshot();
    expect(vm.Script.mock.calls[1][0]).toMatchSnapshot();

    transform('/node_modules/react.js', config);
    // ignores preprocessor
    expect(vm.Script.mock.calls[2][0]).toMatchSnapshot();
  });

  it('reads values from the cache', () => {
    if (skipOnWindows.test()) {
      return;
    }
    const transformConfig = Object.assign(config, {
      transform: [['^.+\\.js$', 'test-preprocessor']],
    });
    transform('/fruits/banana.js', transformConfig);

    const cachePath = getCachePath(mockFs, config);
    expect(fs.writeFileSync).toBeCalled();
    expect(fs.writeFileSync.mock.calls[0][0]).toBe(cachePath);

    // Cache the state in `mockFsCopy`
    const mockFsCopy = mockFs;
    jest.resetModuleRegistry();
    reset();

    // Restore the cached fs
    mockFs = mockFsCopy;
    transform('/fruits/banana.js', transformConfig);

    expect(fs.readFileSync.mock.calls.length).toBe(2);
    expect(fs.readFileSync).toBeCalledWith('/fruits/banana.js', 'utf8');
    expect(fs.readFileSync).toBeCalledWith(cachePath, 'utf8');
    expect(fs.writeFileSync).not.toBeCalled();

    // Don't read from the cache when `config.cache` is false.
    jest.resetModuleRegistry();
    reset();
    mockFs = mockFsCopy;
    transformConfig.cache = false;
    transform('/fruits/banana.js', transformConfig);

    expect(fs.readFileSync.mock.calls.length).toBe(1);
    expect(fs.readFileSync).toBeCalledWith('/fruits/banana.js', 'utf8');
    expect(fs.readFileSync).not.toBeCalledWith(cachePath, 'utf8');
    expect(fs.writeFileSync).toBeCalled();
  });
});
