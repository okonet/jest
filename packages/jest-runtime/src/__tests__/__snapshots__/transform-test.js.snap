exports[`transform transforms a file properly 1`] = `
"({\"Object.<anonymous>\":function(module,exports,require,__dirname,__filename,global,jest){/* istanbul ignore next */var cov_25u22311x4 = function () {var path = \"/fruits/banana.js\",hash = \"04636d4ae73b4b3e24bf6fba39e08c946fd0afb5\",global = new Function(\'return this\')(),gcv = \"__coverage__\",coverageData = { path: \"/fruits/banana.js\", statementMap: { \"0\": { start: { line: 1, column: 0 }, end: { line: 1, column: 26 } } }, fnMap: {}, branchMap: {}, s: { \"0\": 0 }, f: {}, b: {}, _coverageSchema: \"332fd63041d2c1bcb487cc26dd0d5f7d97098a6c\" },coverage = global[gcv] || (global[gcv] = {});if (coverage[path] && coverage[path].hash === hash) {return coverage[path];}coverageData.hash = hash;return coverage[path] = coverageData;}();++cov_25u22311x4.s[0];module.exports = \"banana\";
}});"
`;

exports[`transform transforms a file properly 2`] = `
"({\"Object.<anonymous>\":function(module,exports,require,__dirname,__filename,global,jest){/* istanbul ignore next */var cov_23yvu8etmu = function () {var path = \"/fruits/kiwi.js\",hash = \"fd689b0dc8cb036f0f2381fc209940f4138ade51\",global = new Function(\'return this\')(),gcv = \"__coverage__\",coverageData = { path: \"/fruits/kiwi.js\", statementMap: { \"0\": { start: { line: 1, column: 0 }, end: { line: 1, column: 30 } }, \"1\": { start: { line: 1, column: 23 }, end: { line: 1, column: 29 } } }, fnMap: { \"0\": { name: \"(anonymous_0)\", decl: { start: { line: 1, column: 17 }, end: { line: 1, column: 18 } }, loc: { start: { line: 1, column: 23 }, end: { line: 1, column: 29 } } } }, branchMap: {}, s: { \"0\": 0, \"1\": 0 }, f: { \"0\": 0 }, b: {}, _coverageSchema: \"332fd63041d2c1bcb487cc26dd0d5f7d97098a6c\" },coverage = global[gcv] || (global[gcv] = {});if (coverage[path] && coverage[path].hash === hash) {return coverage[path];}coverageData.hash = hash;return coverage[path] = coverageData;}();++cov_23yvu8etmu.s[0];module.exports = () => {/* istanbul ignore next */++cov_23yvu8etmu.f[0];++cov_23yvu8etmu.s[1];return \"kiwi\";};
}});"
`;

exports[`transform uses the supplied preprocessor 1`] = `
"({\"Object.<anonymous>\":function(module,exports,require,__dirname,__filename,global,jest){
          const TRANSFORMED = {
            filename: \'/fruits/banana.js\',
            script: \'module.exports = \"banana\";\',
            config: \'{\"cache\":true,\"cacheDirectory\":\"/cache/\",\"name\":\"test\",\"preprocessorIgnorePatterns\":[\"/node_modules/\"],\"rootDir\":\"/\",\"scriptPreprocessor\":\"test-preprocessor\"}\',
          };
        
}});"
`;

exports[`transform uses the supplied preprocessor 2`] = `
"({\"Object.<anonymous>\":function(module,exports,require,__dirname,__filename,global,jest){module.exports = \"react\";
}});"
`;
